### DISCUSSION TITLE
ebus Energieintegral

### DISCUSSION THREAD

**Discussion by dabbelju007** [john30/ebusd#740] [Q&A] [answered]
Hallo zusammen,

ich möchte mich vorne weg bei allen Beteiligten, insbesondere natürlich bei @john30, für diese Projekt hier bedanken. Es ist einfach großartig, was hier passiert. Danke dafür!

Jetzt zu meinem Anliegen. Ich habe folgendes Setup (Vaillant):

- aroTHERM plus VWL 125/6
- aroTHERM plus VWZ MEH 97/6
- uniSTOR plus VIH RW 300/3 BR
- sensoCOMFORT VRC 720
- sensoNET VR 921 (nur zum spielen)

ebus läuft einwandfrei und ich kann schon fast alles machen, was ich benötige. Steuerung über Node Red, Daten sammeln in Zabbix.

Die letzten Tage habe ich mich damit beschäftigt, wie ich Daten aus dem Live Monitor der hmu auslesen kann, da mich besonders das Energieintegral interessiert. Dabei bin ich durch den Beitrag #437 (bzw. dem follow up #720) von @Yunkie zu 99% zum Ziel gekommen. Ich kann das Energieintegral auslesen. Allerdings nicht wirklich in "echtzeit" sondern scheinbar nur alle 4,2 Minuten. Ist jetzt nicht dramatsch schlimm, aber es interessiert mich schon warum.

Beim decoden der unbekannten Hexwerte bin ich auf sowas wie 7108b51a04056a3221 / 0a6a082554ffffff000000 gestossen und habe da den vermuteten Wert gefunden und in die lokale CSV eingebaut. Was mir dabei aufgefallen ist, ist das was in #720 von @jkrauss82 beschrieben wurde:

Ich habe an einer Stelle Werte, die scheinbar einen Zeitfaktor beinhalten. Hier ein paar Beispiele:

7108b51a0405**da**3221
7108b51a0405**be**3221
7108b51a0405**ab**3221
7108b51a0405**a2**3221
7108b51a0405**8a**3221
7108b51a0405**86**3221
7108b51a0405**83**3221
7108b51a0405**7a**3221
7108b51a0405**73**3221
7108b51a0405**6a**3221
7108b51a0405**62**3221
7108b51a0405**5e**3221
7108b51a0405**5e**3221
7108b51a0405**33**3221
7108b51a0405**2a**3221
7108b51a0405**12**3221
7108b51a04050**e3**221

Antworten sehen dann zum beispiel so aus: 0a6a08254fffffff000000. Diese habe ich dank des ebusctl grab result decode super "entziffern" können. 54ffff ist das, was mich interessiert. Und wie gesagt auch nach einbau in die csv wie erwartet ausgegeben wird.

Master klar, Slave klar, PBSB klar. Dann kommt immer die Länge 04 und die ID 05 (ich hoffe, das habe ich richtig ausgedrückt). Dann kommt ein Wert, der sich ändert und danach immer 3221. 

Die Werte variieren in Dezimal von 0 bis 254. Durch Beobachtung habe ich die Vermutung, dass es sich dabei um Sekunden handelt. Denn 254 sind in Minuten 4,2 und das entspricht ziemlich genau dem zeitraum, wann neue Werte (bei Änderung) über den Bus kommen.

Hat jemand Erfahrungen damit und kann mir sagen, ob ich das noch besser machen kann?

Danke fürs lesen und viele Grüße
Dabbelju

**Comment by jkrauss82** [comment, marked as answer]
> Allerdings nicht wirklich in "echtzeit" sondern scheinbar nur alle 4,2 Minuten. Ist jetzt nicht dramatsch schlimm, aber es interessiert mich schon warum.

Das hängt mit der Funktionsweise von ebusd zusammen. Intern werden ausgelesene Werte 5 Minuten gecacht. Um einen aktuellen Wert auszulesen, kann man `ebusctl` mit dem Parameter `-f` dazu bringen, immer eine Anfrage an den Bus zu stellen und so den aktuellen Wert zu erhalten.

Vorsicht: zumindest bei mir führt ein zu häufiges Abfragen am Bus zu Fehlern wie "arbitration lost". Habe bei mir einen Cronjob, der gewisse Werte, bspw. die Kompressormodulation, einmal pro Minute abfragt, das funktioniert recht verlässlich. Dann sind sie auch im HTTP Interface aktuell.

**Reply by dabbelju007** [reply]
Hi (ich nehme an) Jonas,

danke für die schnelle Antwort. Werde ich gleich mal ausprobieren, klingt aber nach einem guten grund und einer Lösung.

Aus den anderen Posts habe ich gelernt, dass Du Dich auch sehr gut mit dem Bau von den CSVs auskennst. Darf ich daher fragen, ob die Zeile, die ich für das Auslesen des Integrals gebaut habe so Sinn macht? Funktionieren tut es:

r,,integral,Energieintegral,,08,B51a,05**00**3221,,,,,,,,,integral,,,

Kann ich da das -f mit einbauen? Wenn nicht weiß ich wie ich es mache.

Bei der ID steht ja auch immer der "Zeit" Wert mit drin. Macht das einen Unterschied was ich da eintrage? Also natürlich nur bei dem in diesem Fall 00.

Danke und Gruß
Dabbelju

**Reply by jkrauss82** [reply]
> (ich nehme an) Jonas

Genau :slightly_smiling_face: 

Kenne mich so gut leider nicht aus mit den CSVs, aber wenn das klappt bei Dir sollte es passen. Für die Zeit-Komponente würde ich es einfach mit einem statischen Wert probieren und schauen, ob vom Slave dann jeweils trotzdem richtige Antworten kommen. In dem von mir beobachteten Fall hat das geklappt und dann spricht denke ich nichts dagegen, das so im CSV einzutragen.

Das `-f` ist für den Aufruf von `ebusctl` auf der Kommandozeile gedacht, hat also gar nix mit den CSV-Dateien zu tun. Willst Du bspw. die aktuelle Kompressormodulation vom Bus auslesen, könntest Du das mit folgendem Kommando machen (z.B. wenn Du via SSH auf einem Raspi eingeloggt bist):

`ebusctl read -f CurrentCompressorUtil`

Dadurch wird der Cache aktualisiert und je nachdem wo Du die Werte mit node red ausliest, müssten die dort ebenfalls aktuell sein. Wie geschrieben habe ich bei mir dafür einen Cronjob laufen, der die Kommandos für die Werte die mich interessieren einmal pro Minute mit `-f` anfragt.

**Reply by dabbelju007** [reply]
Perfekt. Ich mache es noch ein wenig anders. Ich nutze Zabbix um Daten zu holen und wegzuschreiben. Da führt man die commandos ebusctl read quasi direkt aus. Es läuft super. Vielen lieben Dank und auf Bald!

**Comment by BenKrae** [comment]
Hallo @dabbelju007,
danke für Deinen Einsatz!
Ich habe nahezu die gleiche Setup wie Du, lediglich die VWL 105 statt der 125.
Ich habe Deine Zeile zum Auslesen des Energieintegrals in der lokalen 08.hmu00.csv ergänzt.
Im HTTP-JSON wird der Parameter auch dargestellt, allerdings erhalte ich konstant den Wert "19493" angezeigt.
Hast Du an Deinem CSV-Eintrag noch etwas verändert?
`r,,integral,Energieintegral,,08,B51a,05003221,,,,,,,,,integral,,,`

**Reply by dabbelju007** [reply]
Hallo BenKrae,

ich habe nichts mehr ergänzt. Wie ist denn Dein Setup bzgl. des ebsu? Ich habe einen esera ebus koppler und das dann per USB an einen Raspberry Pi angeschlossen. Da läuft der ebusd drauf. Meine Fragen an Dich:

1. Was ergibt die Abfrage auf dem Gerät wo der ebusd läuft: ebusctl read integral? Bekommst Du da einen Wert, den halbwegs plausibel ist? Ich habe das am Anfang mit dem Wert aus dem Live Monitor der HMU (Hydraulikstation) verglichen. Ich würde mal tippen, dass der halbwegs stimmt aber mit unter bis zu 5 Minuten alt ist.

2. Was ergibt ein ebusctl read -f integral?

3. Was meinst Du mit HTTP-JSON? Welches System nutzt Du um die Daten zu erfassen?

viele Grüße
Dabbelju

**Reply by dabbelju007** [reply]
Ah, Moment, ich glaube ich weiß was es bei Dir ist. Da wird das Datentemplate integral benutzt. Das musste ich noch in die _template.csv eingebuat werden. Die liegt bei mir im "root folder" wo der Rest der Config liegt. Da muss ein "integral,SIN,,°min,
" rein. habe ich als letzte zeile dazu gepackt.

Einbauen, ebusd neu starten, kurz warten, "ebusctl read integral" ausführen.

Klappts dann?

**Reply by BenKrae** [reply]
Danke für die schnelle Rückmeldung.
Ich habe `integral,SIN,,°min,` in der _template.csv hinten ergänzt, allerdings erhalte ich weiterhin konstant den Wert 19493.
Ich erhalte den Wert sowohl über `ebusctl read -f integral` als auch im JSON über [IP]:[Port]/data.
Weiss jemand wie ich überprüfen kann ob die richtige _template.csv verwendet wird?
Über `ebusctl info` erhalte ich ja "lediglich" die Info über die geladenen CSVs der Parameter.

**Reply by dabbelju007** [reply]
Moin BenKrae,

Du musst ja lokale CSV files nutzen, sonst hättest Du den Wert Integral ja nicht einbauen können, oder? Also Test könntest Du die Zeile mal wie folgt

`r,,integral,Energieintegral,,08,B51a,05003221,,,SIN,,,,,,,,,`

Dadurch unmgest Du die Nutzung des Templates.

Dabbelju

**Reply by dabbelju007** [reply]
@BenKrae: Konntest du das ausprobieren? Und kannst Du bitte mal das Ergebnis von "ebusctl info" posten? Es interessiert mich die Zeile mit der HMU.

Danke und Gruß
Dabbelju

**Reply by BenKrae** [reply]
Hallo @dabbelju007 : Ich habe Deinen Vorschlag getestet, allerdings leider mit dem selben Resultat. Bei mir wird weiterhin konstant der Wert 19493 ausgegeben.

Hier der Auszug aus meinem ebusctl info.
Würde mich sehr freuen wenn Du noch eine Idee hast, wo der Fehler liegen könnte.

> $ ebusctl info
> version: ebusd 22.4.v22.4
> update check: OK, vaillant/08.hmu.csv: different version available
> device: /dev/ttyUSB0, enhanced
> access: *
> signal: acquired
> symbol rate: 22
> max symbol rate: 1314
> min arbitration micros: 319
> max arbitration micros: 9995
> min symbol latency: 0
> max symbol latency: 33
> reconnects: 0
> masters: 5
> messages: 538
> conditional: 0
> poll: 0
> update: 10
> address 00: master #1
> address 03: master #11
> address 04: slave #25, ebusd, scanning
> address 05: slave #1, scanned "MF=Vaillant;ID=VR921;SW=2706;HW=5703"
> address 08: slave #11, scanned "MF=Vaillant;ID=HMU00;SW=0607;HW=5103", loaded "vaillant/08.hmu00.csv"
> address 10: master #2
> address 15: slave #2, scanned "MF=Vaillant;ID=70000;SW=0614;HW=6903", loaded "vaillant/15.700.csv"
> address 26: slave, scanned "MF=Vaillant;ID=VR_71;SW=0104;HW=0503", loaded "vaillant/26.vr_71.csv"
> address 71: master #9
> address 76: slave #9, scanned "MF=Vaillant;ID=VWZIO;SW=0607;HW=5103"
> address ff: master #25, ebusd
>

**Reply by dabbelju007** [reply]
Hi @BenKrae,

ich habe mich mittlerweile mal damit beschäftigt, wie man das besser in die CSV einbaut. Bis dato habe ich einfach nur geschaut, dass es funktioniert. Mittlerweile sieht das in meiner 08.hmu.csv so aus:

`# some more statistics from Live Monitor,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
r,,EnergyIntegral,,,,,21,,,integral,,°min,aktuelles Energieintegral in °min,,,,,,,,,,,,,,,,,,
r,,CurrentYieldPower,,,,,23,,,D1B,10,kW,aktuell gewonnene Energie aus der Umwelt in kW,,,,,,,,,,,,,,,,,,
r,,CurrentConsumedPower,,,,,24,,,D1B,10,kW,aktueller Stromverbrauch der Wärmepumpe in kW,,,,,,,,,,,,,,,,,,
r,,CurrentCompressorUtil,,,,,25,,,D2C,,%,aktuelle Kompressorauslastung in Prozent,,,,,,,,,,,,,,,,,,
`

Ich habe gelernt, dass es auch drauf ankommt an welche Stelle man einen Eintrag in der CSV macht, weil mitunter Werte "vererbt" werden. Deine 08.hmu.csv sollte im Moment so aussehen:

`# some more statistics from Live Monitor,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
r,,CurrentYieldPower,,,,,23,,,D1B,10,kW,aktuell gewonnene Energie aus der Umwelt in kW,,,,,,,,,,,,,,,,,,
r,,CurrentConsumedPower,,,,,24,,,D1B,10,kW,aktueller Stromverbrauch der Wärmepumpe in kW,,,,,,,,,,,,,,,,,,
r,,CurrentCompressorUtil,,,,,25,,,D2C,,%,aktuelle Kompressorauslastung in Prozent,,,,,,,,,,,,,,,,,,
`

Magst Du das mal ausprobieren und sagen ob es dann funktioniert?

**Reply by volvodani** [reply]
Halle @dabbelju007 
Das mit dener Zeile funtkioniert einwandfrei für EnergieIntegral . Ich würde gerne den 
COP 
Kompressorstunden
Und die Soletemperaturen aber das muss man aktiv abfragen ist aber in den CSV noch nicht hinterlegt.
noch rausfinden der ist aber leider nicht hinterlegt.

Meine Konfig ist eine VWF 117/4 mit Tiefenbohrung und nem ESERA Adapter.
Meine Adatpive Heizkurne läuft mittlerweile perfekt.
 Aber ich würde gern noch ein bisschen mehr loggen

Mein Auszug
`pi@raspberrypi:~ $ ebusctl read -f EnergyIntegral
-7
`
`pi@raspberrypi:~ $ ebusctl i
version: ebusd 22.4.v22.4
device: /dev/ttyUSB0
signal: acquired
symbol rate: 23
max symbol rate: 130
min arbitration micros: 325
max arbitration micros: 490
min symbol latency: 0
max symbol latency: 7
reconnects: 0
masters: 5
messages: 489
conditional: 0
poll: 0
update: 10
address 00: master #1
address 03: master #11
address 05: slave #1, scanned "MF=Vaillant;ID=VR921;SW=2706;HW=5703"
address 08: slave #11, scanned "MF=Vaillant;ID=HMU00;SW=0308;HW=0403", loaded "vaillant/08.hmu.csv"
address 10: master #2
address 15: slave #2, scanned "MF=Vaillant;ID=70000;SW=0614;HW=6903", loaded "vaillant/15.700.csv"
address 31: master #8, ebusd
address 36: slave #8, ebusd, scanning
address 71: master #9
address 76: slave #9, scanned "MF=Vaillant;ID=VWZ00;SW=0308;HW=0403"

`

Gruß
DerDani

**Reply by dabbelju007** [reply]
Hi @volvodani,

danke für Dein Feedback. Klappt es bei Dir mit diesem Eintrag hier?

`# some more statistics from Live Monitor,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,, r,,EnergyIntegral,,,,,21,,,integral,,°min,aktuelles Energieintegral in °min,,,,,,,,,,,,,,,,,, r,,CurrentYieldPower,,,,,23,,,D1B,10,kW,aktuell gewonnene Energie aus der Umwelt in kW,,,,,,,,,,,,,,,,,, r,,CurrentConsumedPower,,,,,24,,,D1B,10,kW,aktueller Stromverbrauch der Wärmepumpe in kW,,,,,,,,,,,,,,,,,, r,,CurrentCompressorUtil,,,,,25,,,D2C,,%,aktuelle Kompressorauslastung in Prozent,,,,,,,,,,,,,,,,,, `

Kompressorstunden und COP bin ich selber grade dabei. Bei den Soletemperaturen werde ich nicht helfen können die zu debuggen, weil ich ja eine LWP habe.

viele Grüße
Dabbelju

**Reply by volvodani** [reply]
Genau das meinte ich ich habe die /etc/ebusd/vaillant/08.hmu.csv  angepasst mtid er EngeryIntegral Zeile und es funktionierte auf Anhieb mit read.

Gruß
Daniel 
Bei der LWP müssten die temperauren von der AirUnit aussen  (Vor und Rücklauf) sich denke ich mit der Soletemperatur übereinstimmen.
Sag mal wie machst du das mit dem Rausfinden welche Register du abfragen musst. Ich stehe da voll auf dem Schlauch irgenwei muss manja auf die Werte kommen die man zur Abfrage startet.
Ich würde mich auch selber mal dransetzten an die Pumpe und suchen. 
Ein einfacher Dump des EBUS reicht hier nicht, da die Werte aktiv gefragt werden müssen und nicht auf dem Bus "rumschwirren".
Man müsste einen Anhaltspunkt haben was man da fragen will. Sowas wie ne Liste mit allen Werten die man aktiv fragen kann.
Gruß
Daniel
